content/BCPatterns/generic-method-pattern/index.md

```diff
@@ -0,0 +1,327 @@
+
+## Description
+
+What if we have some kind of "standard way" to always write our code.  The *Generic Method Pattern* is kind of like what it says: a generic way to implement a method.
+
+### What is a method?
+
+Well, a method is *a significant piece of business logic* - may be best explained by some examples:
+
+- Posting a document
+- Any button on a page that executes business logic
+- a batch run to send emails
+- ...
+
+In a way, except "data validation", most of the things we write in our daily life, are methods.
+
+### The Pattern
+
+**One method, one codeunit**
+The idea is to put the code in one *encapsulated* codeunit with the purpose to have all the code in that one codeunit function for that one method.  This way, the codeunit will stay relatively small and readable.
+
+Let me start by showing an example, so you can re:
+
+```AL
+codeunit 53100 "WLD BlockCustomer Meth"
+{
+    internal procedure BlockCustomer(var Cust: Record Customer; HideDialog: Boolean);
+    var
+        Handled: Boolean;
+    begin
+        if not ConfirmBlockCustomer(HideDialog) then exit;
+        OnBeforeBlockCustomer(Cust, Handled);
+        DoBlockCustomer(Cust, Handled);
+        OnAfterBlockCustomer(Cust);
+        AcknowledgeBlockCustomer(HideDialog)
+    end;
+
+    local procedure DoBlockCustomer(var Cust: Record Customer; Handled: Boolean);
+    begin
+        if Handled then
+            exit;
+
+        Cust.Blocked := Cust.Blocked::All;
+        Cust.Modify(true);
+    end;
+
+    local procedure ConfirmBlockCustomer(HideDialog: Boolean): Boolean
+    var
+        ConfirmManagement: Codeunit "Confirm Management";
+        ConfirmQst: Label 'Are you sure?';
+        DefaultAnswer: Boolean;
+    begin
+        DefaultAnswer := true;
+
+        if HideDialog then exit(DefaultAnswer);
+        exit(ConfirmManagement.GetResponseOrDefault(ConfirmQst, DefaultAnswer));
+    end;
+
+    local procedure AcknowledgeBlockCustomer(HideDialog: Boolean)
+    var
+        AcknowledgeMsg: Label 'You successfully executed "BlockCustomer"';
+    begin
+        if not GuiAllowed or HideDialog then exit;
+        Message(AcknowledgeMsg);
+    end;
+
+    [IntegrationEvent(false, false)]
+    local procedure OnBeforeBlockCustomer(var Cust: Record Customer; var Handled: Boolean);
+    begin
+    end;
+
+    [IntegrationEvent(false, false)]
+    local procedure OnAfterBlockCustomer(var Cust: Record Customer);
+    begin
+    end;
+}
+```
+
+Within that codeunit, the pattern is always the same:
+- One public (internal) procedure
+- The rest is always local
+
+So, from outside the codeunit, there is only one clear entrypiont: that one global function with its parameters.
```

Comment:
Typo: entrypoint

Reply:
"that one global function" -> it is an (public) internal function, right? (see two lines above).
So maybe it is better to say "that one internal function".